### DISCUSSION TITLE
java enum의 장점과 단점에 대해 설명 해주세요.

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#53] [Java]
.

**Comment by gyungchan-jo** [comment]
enum 장점:

1. 코드 가독성 향상: 열거형을 사용하면 관련된 상수를 논리적으로 그룹화할 수 있어 코드를 이해하기 쉽게 만들고, 가독성을 향상시킵니다.
2. 타입 안정성(Type Safety): 열거형은 컴파일러가 타입을 검사할 수 있도록 하여 오류를 줄일 수 있습니다.
3. 스레드 안전성(Thread Safety): 열거형은 JVM에 의해 싱글톤 패턴으로 구현되어 있어 멀티스레드 환경에서 안전하게 사용할 수 있습니다.

enum 단점:

1. 상수 개수 제한: 열거형은 미리 정의된 상수 집합을 사용하므로, 동적으로 상수를 추가할 수 없습니다. 
2. 메모리 사용량 증가: 열거형은 클래스로 컴파일되기 때문에 일반적인 상수와 비교하여 메모리 사용량이 더 많을 수 있습니다.
3. 유연성 감소: 열거형은 상속을 지원하지 않습니다. 따라서 열거형을 상속하거나 인터페이스를 구현하는 것은 불가능합니다. 이로 인해 유연성이 제한될 수 있습니다.

**Comment by le2sky** [comment]
Enum을 통해 다음과 같은 장점을 얻을 수 있습니다.

코드가 단순해지고, 가독성이 좋아집니다.
새로운 상수 값이 필요한 경우에도 쉽게 추가 가능합니다.
연관 상수들이 한 곳에서 관리되기 때문에 유지보수에 용이합니다.
자동완성, 오타검증과 같은 IDE의 적극적인 지원을 받을 수 있습니다.
허용 가능한 값들을 제한할 수 있습니다. (type safety)

Enum의 단점은 다음과 같습니다.

런타임에 값의 변경과 추가가 제한됩니다.
일종의 객체이기 때문에 메모리 사용량이 증가할 수 있습니다.

**Reply by le2sky** [reply]
https://www.nextree.co.kr/p11686/

**Comment by GIVEN53** [comment]
상수 그룹화 -> 가독성, 유지보수
컴파일 타임 검사 -> 타입 안정성
싱글톤  패턴 -> 멀티 스레드
런타임 추가, 변경 불가
메모리 사용량 증가
상속 불가 -> Enum을 상속받고 있기 때문
